File: ffanalyzer/ffanalyzer/RKKeySet.m
//
//  RKKeySet.m
//  ffanalyzer
//
//  Created by Peter Kollath on 16/11/14.
//  Copyright (c) 2014 Peter Kollath. All rights reserved.
//

#import "RKKeySet.h"

@implementation RKKeySet

-(id)init
{
    self = [super init];
    if (self) {
        self.nextId = 1;
        self.map = [NSMutableDictionary new];
    }
    return self;
}

-(void)addObject:(NSString *)str
{
    if ([self.map valueForKey:str] == nil)
    {
        [self.map setValue:[NSNumber numberWithInt:self.nextId] forKey:str];
        self.nextId++;
    }
}

-(NSInteger)idForKey:(NSString *)str
{
    NSInteger retVal = self.nextId;
    
    NSNumber * number = [self.map valueForKey:str];
    if (number == nil)
    {
        [self.map setValue:[NSNumber numberWithInt:retVal] forKey:str];
        self.nextId++;
    }
    else
    {
        retVal = [number integerValue];
    }
    
    return retVal;
}

-(NSEnumerator *)keyEnumerator
{
    return self.map.keyEnumerator;
}

@end

File: RKKeySet.py

class RKKeySet:
    def __init__(self):
        self.nextId = 1
        self.map = {}


    def addObject(self,str):
        if str not in self.map:
            self.map[str] = self.nextId
            self.nextId+=1

    def idForKey(self,str):
        retVal = self.nextId
        if str not in self.map:
            self.map[str] = retVal
            self.nextId += 1
        else:
            retVal = self.map[str]
        return retVal

    def keys(self):
        return list(self.map.keys())
